File: packages/sdk-js/src/pulse-buffer.ts
/**
 * Batching for free (AGT-10).
 *
 * Grove caps pulses at one request a second per agent. A PulseBuffer lets an
 * agent pulse as often as it really changes phase: each pulse is stamped with
 * the moment it happened and an event id, queued, and sent in a batch no more
 * than once a second. Nothing a fast agent did is refused for being fast.
 *
 *  - The first pulse after a quiet second goes out immediately; anything that
 *    arrives while the buffer is waiting out the cap rides the next batch.
 *  - A refusal for pace (429) puts the batch back at the front and waits the
 *    server's Retry-After. A network error or 5xx does the same, with backoff.
 *    The event ids make both retries safe: the server reports what already
 *    landed as `duplicate` instead of logging it twice.
 *  - A per-item refusal (bad verb, bad url, a stale timestamp) is final for
 *    that item and is never retried; the rest of the batch still lands.
 *  - Telemetry never breaks a loop: nothing here throws into the caller.
 */
import { GroveApiError } from "./errors.js";
import type { AgentVerb, Presence } from "./types.js";

/** Server limits, mirrored so the buffer never builds a batch the server would refuse whole. */
export const PULSE_BATCH_MAX = 20;
export const PULSE_MIN_INTERVAL_MS = 1000;

/** One pulse in a batch. `at` and `id` are filled in by the buffer when omitted. */
export interface PulseBatchItem {
  verb: AgentVerb;
  detail?: string | null;
  url?: string | null;
  error_text?: string | null;
  /** When it happened. ISO 8601 or epoch ms; no more than 5 minutes ago. */
  at?: string | number;
  /** Your event id (≤64 of A-Za-z0-9._:-). A repeat is reported `duplicate`. */
  id?: string;
  [key: string]: unknown;
}

export interface PulseItemResult {
  index: number;
  id: string | null;
  status: "applied" | "duplicate" | "refused";
  verb: string | null;
  pulsed_at: string | null;
  clamped: boolean;
  code?: "INVALID" | "TIMESTAMP_FUTURE" | "TIMESTAMP_STALE" | string;
  reason?: string;
  [key: string]: unknown;
}

export interface PulseBatchResponse {
  ok: boolean;
  presence: Presence | null;
  results: PulseItemResult[];
  applied: number;
  duplicates: number;
  refused: number;
  [key: string]: unknown;
}

/** What a buffered pulse resolves to once its batch has been answered. */
export interface PulseOutcome {
  result: PulseItemResult;
  /** The body after the batch this pulse rode in. */
  presence: Presence | null;
}

export interface PulseBufferOptions {
  /** Minimum gap between batches. Never below 1000 (the server's cap). */
  intervalMs?: number;
  /** Pulses per batch. 1–20. */
  maxBatch?: number;
  /** Most pulses held while the server is unreachable; the oldest are dropped beyond it. Default 500. */
  maxQueue?: number;
  /** Longest wait between retries after a network error or 5xx. Default 30s. */
  maxBackoffMs?: number;
  onResult?: (response: PulseBatchResponse) => void;
  onError?: (err: unknown) => void;
  onDrop?: (items: PulseBatchItem[]) => void;
}

interface Entry {
  item: PulseBatchItem;
  resolve: (outcome: PulseOutcome | null) => void;
}

function eventId(): string {
  const g = globalThis as { crypto?: { randomUUID?: () => string } };
  if (g.crypto?.randomUUID) return g.crypto.randomUUID();
  return `pulse_${Date.now().toString(36)}_${Math.random().toString(36).slice(2, 12)}`;
}

export class PulseBuffer {
  private queue: Entry[] = [];
  private timer: ReturnType<typeof setTimeout> | null = null;
  private inFlight = false;
  private nextAt = 0;
  private backoffMs = 0;
  private closed = false;
  private idle: Array<() => void> = [];
  private readonly intervalMs: number;
  private readonly maxBatch: number;
  private readonly maxQueue: number;
  private readonly maxBackoffMs: number;

  constructor(
    private readonly send: (items: PulseBatchItem[]) => Promise<PulseBatchResponse>,
    private readonly opts: PulseBufferOptions = {},
  ) {
    this.intervalMs = Math.max(PULSE_MIN_INTERVAL_MS, opts.intervalMs ?? PULSE_MIN_INTERVAL_MS);
    this.maxBatch = Math.min(PULSE_BATCH_MAX, Math.max(1, Math.floor(opts.maxBatch ?? PULSE_BATCH_MAX)));
    this.maxQueue = Math.max(1, opts.maxQueue ?? 500);
    this.maxBackoffMs = Math.max(this.intervalMs, opts.maxBackoffMs ?? 30_000);
  }

  /** Pulses waiting to be sent (not counting a batch in flight). */
  get pending(): number {
    return this.queue.length;
  }

  /**
   * Queue a pulse, stamped now. Resolves when its batch is answered: with the
   * item's result, or null if it was dropped or its batch was refused whole.
   * Never rejects.
   */
  push(
    verb: AgentVerb,
    detail?: string | null,
    extra: { url?: string | null; errorText?: string | null; at?: string | number; id?: string } = {},
  ): Promise<PulseOutcome | null> {
    const item: PulseBatchItem = { verb, at: extra.at ?? new Date().toISOString(), id: extra.id ?? eventId() };
    if (detail != null) item.detail = detail;
    if (extra.url !== undefined) item.url = extra.url;
    if (extra.errorText !== undefined) item.error_text = extra.errorText;
    return this.pushItem(item);
  }

  /** Queue an item you built yourself. Same contract as `push`. */
  pushItem(item: PulseBatchItem): Promise<PulseOutcome | null> {
    if (this.closed) return Promise.resolve(null);
    const stamped: PulseBatchItem = { ...item, at: item.at ?? new Date().toISOString(), id: item.id ?? eventId() };
    return new Promise((resolve) => {
      this.queue.push({ item: stamped, resolve });
      if (this.queue.length > this.maxQueue) {
        const dropped = this.queue.splice(0, this.queue.length - this.maxQueue);
        for (const d of dropped) d.resolve(null);
        this.opts.onDrop?.(dropped.map((d) => d.item));
      }
      this.schedule();
    });
  }

  /** Resolves once everything queued has been answered, still at ≤1 batch a second. */
  flush(): Promise<void> {
    if (!this.queue.length && !this.inFlight) return Promise.resolve();
    return new Promise((resolve) => {
      this.idle.push(resolve);
      this.schedule();
    });
  }

  /** Flush, then refuse new pulses. */
  async close(): Promise<void> {
    await this.flush();
    this.closed = true;
    if (this.timer) clearTimeout(this.timer);
    this.timer = null;
  }

  private schedule(): void {
    if (this.timer || this.inFlight || !this.queue.length) return;
    const wait = Math.max(0, this.nextAt - Date.now());
    this.timer = setTimeout(() => {
      this.timer = null;
      void this.sendOnce();
    }, wait);
    (this.timer as unknown as { unref?: () => void }).unref?.();
  }

  private async sendOnce(): Promise<void> {
    if (this.inFlight || !this.queue.length) return;
    const batch = this.queue.splice(0, this.maxBatch);
    this.inFlight = true;
    this.nextAt = Date.now() + this.intervalMs;
    try {
      const response = await this.send(batch.map((e) => e.item));
      this.backoffMs = 0;
      const byIndex = new Map((response.results ?? []).map((r) => [r.index, r]));
      batch.forEach((entry, i) => {
        const result = byIndex.get(i);
        entry.resolve(result ? { result, presence: response.presence ?? null } : null);
      });
      this.opts.onResult?.(response);
    } catch (err) {
      const apiErr = err instanceof GroveApiError ? err : null;
      if (apiErr?.isRateLimited) {
        // Refused for pace: nothing was written. Back to the front, wait it out.
        this.queue.unshift(...batch);
        const retry = (apiErr.retryAfter ?? 1) * 1000;
        this.nextAt = Date.now() + Math.max(this.intervalMs, retry);
      } else if (!apiErr || apiErr.status >= 500) {
        // Unknown outcome: it may have landed. The ids make resending safe.
        this.queue.unshift(...batch);
        this.backoffMs = Math.min(this.maxBackoffMs, this.backoffMs ? this.backoffMs * 2 : this.intervalMs * 2);
        this.nextAt = Date.now() + this.backoffMs;
        this.opts.onError?.(err);
      } else {
        // Refused whole (UNCLAIMED, not in a room, malformed): resending will not help.
        for (const e of batch) e.resolve(null);
        this.opts.onError?.(err);
      }
      if (this.queue.length > this.maxQueue) {
        const dropped = this.queue.splice(0, this.queue.length - this.maxQueue);
        for (const d of dropped) d.resolve(null);
        this.opts.onDrop?.(dropped.map((d) => d.item));
      }
    } finally {
      this.inFlight = false;
      if (!this.queue.length) {
        const waiters = this.idle.splice(0);
        for (const w of waiters) w();
      }
      this.schedule();
    }
  }
}

File: packages/sdk-py/grove_sdk/pulse_buffer.py
"""Batching for free (AGT-10).

Grove caps pulses at one request a second per agent. A :class:`PulseBuffer`
lets an agent pulse as often as it really changes phase: each pulse is stamped
with the moment it happened and an event id, queued, and sent in a batch no
more than once a second. Nothing a fast agent did is refused for being fast.

- The first pulse after a quiet second goes out immediately; anything that
  arrives while the buffer waits out the cap rides the next batch.
- A refusal for pace (429) puts the batch back at the front and waits the
  server's Retry-After. A network error or 5xx does the same, with backoff.
  The event ids make both retries safe: the server reports what already landed
  as ``duplicate`` instead of logging it twice.
- A per-item refusal (bad verb, bad url, a stale timestamp) is final for that
  item and never retried; the rest of the batch still lands.
- Telemetry never breaks a loop: nothing here raises into the caller.
"""

from __future__ import annotations

import threading
import time
from collections import deque
from datetime import datetime, timezone
from typing import Any, Callable, Deque, Dict, List, Optional
from uuid import uuid4

from .errors import GroveError

#: Server limits, mirrored so the buffer never builds a batch refused whole.
PULSE_BATCH_MAX = 20
PULSE_MIN_INTERVAL = 1.0


def _iso_ms(epoch_seconds: float) -> str:
    stamp = datetime.fromtimestamp(epoch_seconds, tz=timezone.utc)
    return stamp.strftime("%Y-%m-%dT%H:%M:%S.") + "%03dZ" % (stamp.microsecond // 1000)


class PulseBuffer:
    """Queue pulses; send at most one batch (<= 20) a second.

    >>> buffer = grove.pulse_buffer()            # doctest: +SKIP
    >>> buffer.push("tool", "pytest -q")         # doctest: +SKIP
    >>> buffer.close()                           # doctest: +SKIP

    ``send`` takes a list of pulse dicts and returns the batch response.
    ``clock`` (monotonic, for pacing) and ``wall`` (epoch seconds, for ``at``)
    are injectable for tests; so is ``autostart=False``, which runs no thread
    and lets you drive :meth:`pump` yourself.
    """

    def __init__(
        self,
        send: Callable[[List[Dict[str, Any]]], Any],
        interval: float = PULSE_MIN_INTERVAL,
        max_batch: int = PULSE_BATCH_MAX,
        max_queue: int = 500,
        max_backoff: float = 30.0,
        on_result: Optional[Callable[[Any], None]] = None,
        on_error: Optional[Callable[[BaseException], None]] = None,
        on_drop: Optional[Callable[[List[Dict[str, Any]]], None]] = None,
        clock: Callable[[], float] = time.monotonic,
        wall: Callable[[], float] = time.time,
        autostart: bool = True,
    ) -> None:
        self._send = send
        self.interval = max(PULSE_MIN_INTERVAL, float(interval))
        self.max_batch = min(PULSE_BATCH_MAX, max(1, int(max_batch)))
        self.max_queue = max(1, int(max_queue))
        self.max_backoff = max(self.interval, float(max_backoff))
        self._on_result = on_result
        self._on_error = on_error
        self._on_drop = on_drop
        self._clock = clock
        self._wall = wall
        self._queue: Deque[Dict[str, Any]] = deque()
        self._cond = threading.Condition()
        self._in_flight = False
        self._next_at = float("-inf")
        self._backoff = 0.0
        self._closed = False
        self._thread: Optional[threading.Thread] = None
        if autostart:
            self._thread = threading.Thread(target=self._run, name="grove-pulse-buffer", daemon=True)
            self._thread.start()

    @property
    def pending(self) -> int:
        """Pulses waiting to be sent (not counting a batch in flight)."""
        with self._cond:
            return len(self._queue)

    def push(
        self,
        verb: str,
        detail: Optional[str] = None,
        url: Optional[str] = None,
        error_text: Optional[str] = None,
        at: Optional[str] = None,
        id: Optional[str] = None,  # noqa: A002 - the wire name
    ) -> Optional[Dict[str, Any]]:
        """Queue a pulse stamped now. Returns the queued item (with its ``id``), or None once closed."""
        item: Dict[str, Any] = {"verb": verb}
        if detail is not None:
            item["detail"] = detail
        if url is not None:
            item["url"] = url
        if error_text is not None:
            item["error_text"] = error_text
        if at is not None:
            item["at"] = at
        if id is not None:
            item["id"] = id
        return self.push_item(item)

    def push_item(self, item: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Queue a pulse dict you built yourself. ``at`` and ``id`` are filled in when missing."""
        if self._closed:
            return None
        stamped = dict(item)
        stamped.setdefault("at", _iso_ms(self._wall()))
        stamped.setdefault("id", str(uuid4()))
        dropped: List[Dict[str, Any]] = []
        with self._cond:
            self._queue.append(stamped)
            while len(self._queue) > self.max_queue:
                dropped.append(self._queue.popleft())
            self._cond.notify_all()
        if dropped and self._on_drop:
            self._on_drop(dropped)
        return stamped

    def pump(self) -> Optional[float]:
        """Send one batch if one is due. Returns seconds until the next is due, or None when empty."""
        with self._cond:
            if not self._queue:
                return None
            if self._in_flight:
                return self.interval
            wait = self._next_at - self._clock()
            if wait > 0:
                return wait
            batch = [self._queue.popleft() for _ in range(min(self.max_batch, len(self._queue)))]
            self._in_flight = True
            self._next_at = self._clock() + self.interval
        try:
            response = self._send(batch)
            self._backoff = 0.0
            if self._on_result:
                self._on_result(response)
        except GroveError as err:
            if err.is_rate_limited:
                # Refused for pace: nothing was written. Back to the front, wait it out.
                self._requeue(batch, max(self.interval, float(err.retry_after or 1)))
            elif not err.status or err.status >= 500:
                self._retry_later(batch, err)
            else:
                # Refused whole (unclaimed, not in a room, malformed): resending will not help.
                if self._on_error:
                    self._on_error(err)
        except Exception as err:  # noqa: BLE001 - network trouble must never reach the loop
            self._retry_later(batch, err)
        finally:
            with self._cond:
                self._in_flight = False
                self._cond.notify_all()
        with self._cond:
            if not self._queue:
                return None
            return max(0.0, self._next_at - self._clock())

    def flush(self, timeout: Optional[float] = None) -> bool:
        """Block until everything queued has been sent, still at <= 1 batch a second.

        Returns False if ``timeout`` (seconds) ran out first.
        """
        deadline = None if timeout is None else self._clock() + timeout
        if self._thread is None:
            while True:
                wait = self.pump()
                if wait is None and not self._in_flight:
                    return True
                if deadline is not None and self._clock() >= deadline:
                    return False
                time.sleep(min(wait or 0.01, 0.25))
        with self._cond:
            while self._queue or self._in_flight:
                remaining = None if deadline is None else deadline - self._clock()
                if remaining is not None and remaining <= 0:
                    return False
                self._cond.wait(remaining if remaining is not None else 0.25)
            return True

    def close(self, timeout: Optional[float] = None) -> bool:
        """Flush, then stop the thread and refuse new pulses."""
        flushed = self.flush(timeout)
        with self._cond:
            self._closed = True
            self._cond.notify_all()
        return flushed

    # -- internals --------------------------------------------------------

    def _requeue(self, batch: List[Dict[str, Any]], delay: float) -> None:
        dropped: List[Dict[str, Any]] = []
        with self._cond:
            self._queue.extendleft(reversed(batch))
            self._next_at = self._clock() + delay
            while len(self._queue) > self.max_queue:
                dropped.append(self._queue.popleft())
        if dropped and self._on_drop:
            self._on_drop(dropped)

    def _retry_later(self, batch: List[Dict[str, Any]], err: BaseException) -> None:
        # Unknown outcome: it may have landed. The ids make resending safe.
        self._backoff = min(self.max_backoff, self._backoff * 2 if self._backoff else self.interval * 2)
        self._requeue(batch, self._backoff)
        if self._on_error:
            self._on_error(err)

    def _run(self) -> None:
        while True:
            with self._cond:
                while not self._queue and not self._closed:
                    self._cond.wait()
                if self._closed and not self._queue:
                    return
            wait = self.pump()
            if wait:
                with self._cond:
                    self._cond.wait(wait)


__all__ = ["PulseBuffer", "PULSE_BATCH_MAX", "PULSE_MIN_INTERVAL"]
